rust_snuba/rust_arroyo/src/types/mod.rs

```diff
@@ -35,36 +36,125 @@ pub enum TopicOrPartition {
 }
 
 #[derive(Clone, Debug, PartialEq)]
-pub struct Message<T: Clone> {
+pub struct BrokerMessage<T: Clone> {
+    pub payload: T,
     pub partition: Partition,
     pub offset: u64,
-    pub payload: T,
     pub timestamp: DateTime<Utc>,
 }
 
-impl<T: Clone> Message<T> {
-    pub fn new(partition: Partition, offset: u64, payload: T, timestamp: DateTime<Utc>) -> Self {
+impl<T: Clone>BrokerMessage<T> {
+    pub fn new(payload: T, partition: Partition, offset: u64, timestamp: DateTime<Utc>) -> Self {
         Self {
+            payload,
             partition,
             offset,
-            payload,
             timestamp,
         }
     }
-    pub fn next_offset(&self) -> u64 {
-        self.offset + 1
+
+    pub fn replace<TReplaced: Clone>(&self, replacement: TReplaced) -> BrokerMessage<TReplaced> {
+        BrokerMessage {
+            payload: replacement,
+            partition: self.partition.clone(),
+            offset: self.offset,
+            timestamp: self.timestamp,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AnyMessage<T: Clone> {
+    pub payload: T,
+    pub committable: BTreeMap<Partition, u64>
+}
+
+impl<T: Clone> AnyMessage<T> {
+    pub fn new(payload: T, committable: BTreeMap<Partition, u64>) -> Self {
+        Self {
+            payload, committable
+        }
+    }
+
+    pub fn replace<TReplaced: Clone>(&self, replacement: TReplaced) -> AnyMessage<TReplaced>{
+        // TODO: Get rid of the clone
+        AnyMessage{payload: replacement, committable: self.committable.clone()}
+    }
+}
+
+
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum InnerMessage<T: Clone> {
+    BrokerMessage(BrokerMessage<T>),
+    AnyMessage(AnyMessage<T>),
+}
+
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Message<T: Clone> {
+    pub inner_message:  InnerMessage<T>,
+}
+
+impl<T: Clone> Message<T> {
+    pub fn payload(&self) -> T {
+        match &self.inner_message {
+            InnerMessage::BrokerMessage(BrokerMessage{payload, ..}) => payload.clone(),
+            InnerMessage::AnyMessage(AnyMessage{payload, ..}) => payload.clone(),
+        }
+    }
+
+    pub fn committable(&self) -> BTreeMap<Partition, u64> {
+        match &self.inner_message {
+            InnerMessage::BrokerMessage(BrokerMessage{partition, offset, ..}) => {
+                let mut map = BTreeMap::new();
+                // TODO: Get rid of the clone
```

Comment:
Partition IMO should be Copy, but you can't do that because the topic is contained as a String. You could make it so that the topic names are statically allocated somewhere (easily achievable using `Box::leak`), and then all the Partitions can hold a `&'static str` for the topic name which is just a pointer to some string that Rust now knows will live forever